File: infrastructures/ddd/python/src/domain/models/example_model.py
"""
Example Business Model Object (BMO) demonstrating domain layer patterns.
This class:
- Lives in /domain/models/
- Has Model or BMO suffix
- Contains business logic and invariants
- MUST NOT have database/ORM attributes
- Uses member field naming convention: _variable
"""

from datetime import datetime
from typing import Optional
from domain.models.infra import BaseModel


class ExampleModel(BaseModel):
    """
    Example domain model demonstrating DDD patterns.
    
    This model:
    - Encapsulates business rules
    - Enforces invariants through methods
    - Uses factory methods for creation
    - Provides hydration for persistence layer
    """
    
    def __init__(self):
        """Private constructor. Use factory methods instead."""
        super().__init__()
        self._name: str = ""
        self._description: str = ""
        self._owner_id: str = ""
        self._is_active: bool = True
    
    @property
    def name(self) -> str:
        """Get the example name."""
        return self._name
    
    @property
    def description(self) -> str:
        """Get the example description."""
        return self._description
    
    @property
    def owner_id(self) -> str:
        """Get the owner user ID."""
        return self._owner_id
    
    @property
    def is_active(self) -> bool:
        """Get the active status."""
        return self._is_active
    
    @staticmethod
    def create(p_name: str, p_description: str, p_owner_id: str) -> 'ExampleModel':
        """
        Factory method for creating new instances.
        
        Args:
            p_name: The name of the example
            p_description: The description of the example
            p_owner_id: The ID of the owner user
            
        Returns:
            A validated ExampleModel instance
            
        Raises:
            ValueError: If validation fails
        """
        model = ExampleModel()
        model._name = p_name
        model._description = p_description
        model._owner_id = p_owner_id
        model._is_active = True
        
        model.validate()
        return model
    
    @staticmethod
    def hydrate(
        p_id: str,
        p_name: str,
        p_description: str,
        p_owner_id: str,
        p_is_active: bool,
        p_created_at: datetime,
        p_updated_at: datetime
    ) -> 'ExampleModel':
        """
        Factory method for hydration from persistence layer.
        
        Args:
            p_id: The unique identifier
            p_name: The name
            p_description: The description
            p_owner_id: The owner ID
            p_is_active: The active status
            p_created_at: The creation timestamp
            p_updated_at: The update timestamp
            
        Returns:
            A hydrated ExampleModel instance
        """
        model = ExampleModel()
        model._id = p_id
        model._name = p_name
        model._description = p_description
        model._owner_id = p_owner_id
        model._is_active = p_is_active
        model._created_at = p_created_at
        model._updated_at = p_updated_at
        
        return model
    
    def update_details(self, p_name: str, p_description: str) -> None:
        """
        Updates the example details.
        
        Args:
            p_name: The new name
            p_description: The new description
            
        Raises:
            ValueError: If validation fails
        """
        if not p_name or not p_name.strip():
            raise ValueError("Name cannot be empty")
        
        self._name = p_name
        self._description = p_description
        self._updated_at = datetime.utcnow()
        
        self.validate()
    
    def activate(self) -> None:
        """Activates the example."""
        self._is_active = True
        self._updated_at = datetime.utcnow()
    
    def deactivate(self) -> None:
        """Deactivates the example."""
        self._is_active = False
        self._updated_at = datetime.utcnow()
    
    def validate_ownership(self, p_user_id: str) -> None:
        """
        Validates that the user owns this resource.
        
        Args:
            p_user_id: The user ID to validate
            
        Raises:
            PermissionError: If the user does not own the resource
        """
        if self._owner_id != p_user_id:
            raise PermissionError("User does not own this resource")
    
    def validate(self) -> None:
        """
        Validates business invariants.
        
        Raises:
            ValueError: If validation fails
        """
        super().validate()
        
        if not self._name or not self._name.strip():
            raise ValueError("Example name cannot be empty")
        
        if len(self._name) > 100:
            raise ValueError("Example name cannot exceed 100 characters")
        
        if not self._owner_id or not self._owner_id.strip():
            raise ValueError("Example must have an owner")

File: infrastructures/ddd/csharp/src/Domain/Models/ExampleModel.cs
using System;

namespace OddlyDdd.Domain.Models
{
    /// <summary>
    /// Example Business Model Object (BMO) demonstrating domain layer patterns.
    /// This class:
    /// - Lives in /Domain/Models/
    /// - Has Model or BMO suffix
    /// - Contains business logic and invariants
    /// - MUST NOT have database/ORM attributes
    /// - Uses member field naming convention: _variable
    /// </summary>
    public class ExampleModel : BaseModel
    {
        private string _name = string.Empty;
        private string _description = string.Empty;
        private string _ownerId = string.Empty;
        private bool _isActive;

        public string Name
        {
            get => _name;
            private set => _name = value;
        }

        public string Description
        {
            get => _description;
            private set => _description = value;
        }

        public string OwnerId
        {
            get => _ownerId;
            private set => _ownerId = value;
        }

        public bool IsActive
        {
            get => _isActive;
            private set => _isActive = value;
        }

        // Private constructor for hydration from database
        private ExampleModel()
        {
        }

        // Public factory method for creating new instances
        public static ExampleModel Create(string p_name, string p_description, string p_ownerId)
        {
            var model = new ExampleModel
            {
                _name = p_name,
                _description = p_description,
                _ownerId = p_ownerId,
                _isActive = true
            };

            model.Validate();
            return model;
        }

        // Factory method for hydration from persistence layer
        public static ExampleModel Hydrate(
            string p_id,
            string p_name,
            string p_description,
            string p_ownerId,
            bool p_isActive,
            DateTime p_createdAt,
            DateTime p_updatedAt)
        {
            return new ExampleModel
            {
                _id = p_id,
                _name = p_name,
                _description = p_description,
                _ownerId = p_ownerId,
                _isActive = p_isActive,
                _createdAt = p_createdAt,
                _updatedAt = p_updatedAt
            };
        }

        // Business logic methods
        public void UpdateDetails(string p_name, string p_description)
        {
            if (string.IsNullOrWhiteSpace(p_name))
            {
                throw new ArgumentException("Name cannot be empty", nameof(p_name));
            }

            _name = p_name;
            _description = p_description;
            _updatedAt = DateTime.UtcNow;
            
            Validate();
        }

        public void Activate()
        {
            _isActive = true;
            _updatedAt = DateTime.UtcNow;
        }

        public void Deactivate()
        {
            _isActive = false;
            _updatedAt = DateTime.UtcNow;
        }

        public void ValidateOwnership(string p_userId)
        {
            if (_ownerId != p_userId)
            {
                throw new UnauthorizedAccessException("User does not own this resource");
            }
        }

        // Validate business invariants
        public override void Validate()
        {
            base.Validate();

            if (string.IsNullOrWhiteSpace(_name))
            {
                throw new InvalidOperationException("Example name cannot be empty");
            }

            if (_name.Length > 100)
            {
                throw new InvalidOperationException("Example name cannot exceed 100 characters");
            }

            if (string.IsNullOrWhiteSpace(_ownerId))
            {
                throw new InvalidOperationException("Example must have an owner");
            }
        }
    }
}
